### PR TITLE
grep functions refactoring

### PR DESCRIPTION
**Pull request by qrux0**
Moved some grep logic from libr/core/cmd.c into libr/cons/grep.c (as requested in https://github.com/radare/radare2/pull/7669),
few libr/include/r_cons.h style fixes (removed 	superfluous space in function definitions).

### CODE REVIEW

**Comment by qrux0** [comment]
I will add tests into radare2-regressions soon.

File: libr/cons/grep.c [hunk 1]
```diff
@@ -278,6 +278,40 @@ R_API void r_cons_grep(const char *str) {
 	}
 }
 
+R_API void r_cons_grep_parsecmd(char *cmd, const char *quotestr) {
+	char *ptr;
+	int len = strlen (cmd);
+
+	//  search and modify '~str~?' into '~?str'
+	if (len > 4 && r_str_endswith (cmd, "~?") && cmd[len - 3] != '\\') {
```
Comment (radare) [inline libr/cons/grep.c]: this solution only catches this case.. but in fact what ~...~... should concatenate more than one grep after the other. so applying ~str~?  will result in grepping for the lines matching str and then counting them.

File: libr/cons/grep.c [hunk 2]
```diff
@@ -278,6 +278,40 @@ R_API void r_cons_grep(const char *str) {
 	}
 }
 
+R_API void r_cons_grep_parsecmd(char *cmd, const char *quotestr) {
+	char *ptr;
+	int len = strlen (cmd);
```
Comment (radare) [inline libr/cons/grep.c]: add nullptr checks for cmd and quotestr

File: libr/cons/grep.c (hunk 1 above)
Comment (qrux0) [inline libr/cons/grep.c]: Markdown at github.com messes part of your commentary.
Just to be clear, did you mean that 

    ~str1~str2~?

should do the same as

    ~?str1str2

I.e. r2 need to handle every unescaped '~' from command string, and build one grep expression from multiple intgrep subcommands?

**Comment by radare** [comment]
> On 4 Jun 2017, at 11:14, Andry <notifications@github.com> wrote:
> 
> @qrux0 commented on this pull request.
> 
> In libr/cons/grep.c <https://github.com/radare/radare2/pull/7672#discussion_r120004803>:
> 
> > @@ -278,6 +278,40 @@ R_API void r_cons_grep(const char *str) {
>  	}
>  }
>  
> +R_API void r_cons_grep_parsecmd(char *cmd, const char *quotestr) {
> +	char *ptr;
> +	int len = strlen (cmd);
> +
> +	//  search and modify '~str~?' into '~?str'
> +	if (len > 4 && r_str_endswith (cmd, "~?") && cmd[len - 3] != '\\') {
> Markdown at github.com messes part of your commentary.
> Just to be clear, did you mean that
> 
> ~str1~str2~?
> should do the same as
> 
> ~?str1str2
no, thats not the same thing, ~str1~str2~? is the equivalent to do this:

| grep str1 | grep str2 | wc -l

which in current consgrep is:

~?&str1,str2

but ~str1~str2~? is more readable than ~?&str,str2 :P

you can copmress the expression i guess but there must be some escaping problemsin some situations, but this should be another thing to be tested, can u do the tests for those cases too? thanks!

**Comment by radare** [comment]
ping

**Comment by qrux0** [comment]
Pong.

Thank you, now I get it. I will try to implement this functionality & tests in my spare time. I'm sorry for such slow replies and code writing)

**Comment by radare** [comment]
np :) we are all busy, keep up the good work!

File: libr/cons/grep.c [hunk 3]
```diff
@@ -278,6 +278,94 @@ R_API void r_cons_grep(const char *str) {
...
+		p1[len - 2] = '\0';
+		ns = r_str_append (ns, "?");
+	}
+
+	*p1 = '\0'; // remove grep part from cmd
+
+	i = 0;
+	strsep = "&";
+	// parse words between '~'
+	while ((p2 = find_next_intgrep (p1 + 1, quotes)) != NULL) {
```
Comment (radare) [inline libr/cons/grep.c]: != NULL is not necessary

File: libr/cons/grep.c [hunk 4]
```diff
@@ -278,6 +278,94 @@ R_API void r_cons_grep(const char *str) {
...
+	char *strsep;
+	int len;
+	int i;
+
+	p1 = find_next_intgrep (cmd, quotes);
+	if (p1 == NULL) {
+		return NULL;
+	}
+
+	ns = NULL;
```
Comment (radare) [inline libr/cons/grep.c]: define and initialyze the vars in the same place char *ns = NULL;

File: libr/cons/grep.c [hunk 5]
```diff
@@ -278,6 +278,94 @@ R_API void r_cons_grep(const char *str) {
...
+ * converts "~str1~str2~str3~?" into "?&str1,str2,str3"
+ */
+static char *preprocess_filter_expr(char *cmd, const char *quotes) {
+	char *p1, *p2, *ns;
+	char *strsep;
+	int len;
+	int i;
+
+	p1 = find_next_intgrep (cmd, quotes);
+	if (p1 == NULL) {
```
Comment (radare) [inline libr/cons/grep.c]: juse (!p1) { ...

File: libr/cons/grep.c [hunk 6]
```diff
@@ -278,6 +278,94 @@ R_API void r_cons_grep(const char *str) {
...
+ * Removes grep part from *cmd* and returns newly allocated string
+ * with reshaped grep expression.
+ *
+ * Function converts multiple twiddle expressions into internal representation.
+ * For example:
+ * converts "~str1~str2~str3~?" into "?&str1,str2,str3"
+ */
+static char *preprocess_filter_expr(char *cmd, const char *quotes) {
+	char *p1, *p2, *ns;
+	char *strsep;
```
Comment (radare) [inline libr/cons/grep.c]: this is const char * and should be probably initialized first

File: libr/cons/grep.c [hunk 7]
```diff
@@ -278,6 +278,94 @@ R_API void r_cons_grep(const char *str) {
...
+	return ns;
+}
+
+R_API void r_cons_grep_parsecmd(char *cmd, const char *quotestr) {
+	char *ptr;
+
+	if (!cmd)
+		return;
+
+	ptr = preprocess_filter_expr (cmd, quotestr);
```
Comment (radare) [inline libr/cons/grep.c]: char *ptr = ... and save 2 lines while making it more readable

File: libr/cons/grep.c [hunk 8]
```diff
@@ -278,6 +278,94 @@ R_API void r_cons_grep(const char *str) {
...
+
+	ns = r_str_append (ns, p1 + 1);
+
+	return ns;
+}
+
+R_API void r_cons_grep_parsecmd(char *cmd, const char *quotestr) {
+	char *ptr;
+
+	if (!cmd)
```
Comment (radare) [inline libr/cons/grep.c]: add braces

File: libr/util/str.c [hunk 9]
```diff
@@ -1807,11 +1807,40 @@ R_API void r_str_argv_free(char **argv) {
...
+	}
+	if (strlen (but) >= bsz) {
+		eprintf ("r_str_firstbut: but string too long\n");
+		return NULL;
+	}
+	for (p = s; *p; p++) {
+		isbut = strchr (but, *p);
+		if (isbut) {
+			idx = (int)(size_t)(isbut - but);
+			_b = R_BIT_CHK (b, idx)?
```
Comment (radare) [inline libr/util/str.c]: what about R_BIT_SWAP?

**Comment by radare** [comment]
Can you add some tests for this? i will merge after you fix the comments. Thanks!

**Comment by qrux0** [comment]
About you comment

    char *ptr = ... and save 2 lines while making it more readable

at 360 line.

I think we have to check cmd == NULL before calling *preprocess_filter_expr*, thats why in this case we souldn't initialize **ptr** on definition.

**Comment by radare** [comment]
Please rebase

**Comment by qrux0** [comment]
Done

**Comment by radare** [comment]
thanks!

**Pull request merged**